Implement lastLineBaseline for tables.

sammygill opened the pull request:
#### bf33593289e7c50abaf636fdca635f3bac442eb4
<pre>
Implement lastLineBaseline for tables.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244858">https://bugs.webkit.org/show_bug.cgi?id=244858</a>
rdar://99621230

Reviewed by Alan Bujtas.

CSS Box Alignment Module Level 3 provides a definition of the last
baseline for tables and table rows. We currently keep track of the
baselines for each row if there are cells within that are baseline
aligned, so we just need to add the logic for finding the last baseline
if there are no cells being baseline aligned. This is the same logic
that was already in firstLineBaseline, but instead we are looking at
only the last row. This logic has been refactored out into
baselineFromCellContentEdges and now takes into consideration which
type of baseline alignment we are doing.

Spec reference: <a href="https://www.w3.org/TR/css-align-3/#baseline-export">https://www.w3.org/TR/css-align-3/#baseline-export</a>

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-003-expected.txt:
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::bottomNonEmptySection const):
(WebCore::RenderTable::lastLineBaseline const):
* Source/WebCore/rendering/RenderTable.h:
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::firstLineBaseline const):
(WebCore::RenderTableSection::lastLineBaseline const):
(WebCore::RenderTableSection::baselineFromCellContentEdges const):
* Source/WebCore/rendering/RenderTableSection.h:

Canonical link: <a href="https://commits.webkit.org/255546@main">https://commits.webkit.org/255546@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff0a8496adea8204a217701e5487801b1d9a128c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92804 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2018 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23386 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102533 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162834 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96806 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2018 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30359 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85204 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98687 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98467 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/1371 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79292 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28286 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83284 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82996 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/71403 "Found 2 new API test failures: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36762 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/16911 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34563 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18099 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3841 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38432 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40703 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40347 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37269 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1541,6 +1549,22 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
+
+    const RenderTableSection* tableSection = bottomNonEmptySection();
+    if (!tableSection)
+        return std::optional<LayoutUnit>();
```
return { }

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1541,6 +1549,22 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
+
+    recalcSectionsIfNeeded();
+
+    const RenderTableSection* tableSection = bottomNonEmptySection();
```
auto

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1541,6 +1549,22 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
+    if (!tableSection)
+        return std::optional<LayoutUnit>();
+
+    if (std::optional<LayoutUnit> baseline = tableSection->lastLineBaseline())
```
auto

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1541,6 +1549,22 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
+        return std::optional<LayoutUnit>();
+
+    if (std::optional<LayoutUnit> baseline = tableSection->lastLineBaseline())
+        return (std::optional<LayoutUnit>(baseline.value() + tableSection->logicalTop()));
```
return LayoutUnit { baseline.value() + tableSection->logicalTop() } ?

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -878,18 +878,36 @@ std::optional<LayoutUnit> RenderTableSection::firstLineBaseline() const
+std::optional<LayoutUnit> RenderTableSection::lastLineBaseline() const
+{
+    if (!m_grid.size())
+        return std::optional<LayoutUnit>();
```
return { }

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -878,18 +878,36 @@ std::optional<LayoutUnit> RenderTableSection::firstLineBaseline() const
+        return std::optional<LayoutUnit>();
+    
+    LayoutUnit lastLineBaseline = m_grid[m_grid.size() - 1].baseline;
+    if (lastLineBaseline)
```
if (auto  lastLineBaseline = m_grid[m_grid.size() - 1].baseline)

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1541,6 +1549,22 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
+std::optional<LayoutUnit> RenderTable::lastLineBaseline() const
+{
+    if (isWritingModeRoot() || shouldApplyLayoutContainment())
+        return std::optional<LayoutUnit>();
```
return { }

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1541,6 +1549,22 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
+    if (isWritingModeRoot() || shouldApplyLayoutContainment())
+        return std::optional<LayoutUnit>();
+
+    recalcSectionsIfNeeded();
```
shouldn't lastBaseline be called when we are already done with laying out the section? I am just curious of why this call is needed here.

sammygill replied:
I wasn't entirely sure about that myself so I just decided to keep the logic similar to what was already in firstLineBaseline. I see that `setNeedsSectionRecalc` gets called in a couple different places like `removeColumn`, which gets called from `willBeRemovedFromTree`. I'm not familiar with the scenarios in which can occur. I can remove it if you would like or we can just keep it to remain consistent for now.

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -901,6 +901,14 @@ RenderTableSection* RenderTable::topNonEmptySection() const
 
+RenderTableSection* RenderTable::bottomNonEmptySection() const
+{
+    RenderTableSection* section = bottomSection();
```
auto

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1541,6 +1549,22 @@ std::optional<LayoutUnit> RenderTable::firstLineBaseline() const
+
+    if (std::optional<LayoutUnit> baseline = tableSection->lastLineBaseline())
+        return (std::optional<LayoutUnit>(baseline.value() + tableSection->logicalTop()));
+    return std::optional<LayoutUnit>();
```
return { }

sammygill replied:
Changed

alanbaradlay commented on Source/WebCore/rendering/RenderTableSection.cpp:
```diff
@@ -878,18 +878,36 @@ std::optional<LayoutUnit> RenderTableSection::firstLineBaseline() const
+std::optional<LayoutUnit> RenderTableSection::baselineFromCellContentEdges(ItemPosition alignment) const
+{
+    ASSERT(alignment == ItemPosition::Baseline || alignment == ItemPosition::LastBaseline);
+    const Row& row = alignment == ItemPosition::Baseline ? m_grid[0].row : m_grid[m_grid.size() - 1].row;
```
auto

sammygill replied:
Changed

sammygill committed:
Implement lastLineBaseline for tables.
https://bugs.webkit.org/show_bug.cgi?id=244858
rdar://99621230

Reviewed by Alan Bujtas.

CSS Box Alignment Module Level 3 provides a definition of the last
baseline for tables and table rows. We currently keep track of the
baselines for each row if there are cells within that are baseline
aligned, so we just need to add the logic for finding the last baseline
if there are no cells being baseline aligned. This is the same logic
that was already in firstLineBaseline, but instead we are looking at
only the last row. This logic has been refactored out into
baselineFromCellContentEdges and now takes into consideration which
type of baseline alignment we are doing.

Spec reference: https://www.w3.org/TR/css-align-3/#baseline-export

* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/alignment/flex-align-baseline-table-003-expected.txt:
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::bottomNonEmptySection const):
(WebCore::RenderTable::lastLineBaseline const):
* Source/WebCore/rendering/RenderTable.h:
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::firstLineBaseline const):
(WebCore::RenderTableSection::lastLineBaseline const):
(WebCore::RenderTableSection::baselineFromCellContentEdges const):
* Source/WebCore/rendering/RenderTableSection.h:

Canonical link: https://commits.webkit.org/255546@main
